### DISCUSSION TITLE
MQTT Retain Flag

### DISCUSSION THREAD

**Discussion by Grillfreund** [jomjol/AI-on-the-edge-device#708] [Ideas]
Hello jomjol,

first I would like to thank you for this very cool project. I printed a case for my gas meter during weekend. Today everything is working fine. I cant believe how powerful this lite piece of software is! Again: Thank you!

But now a little feature request (probably it is already possible but I didn't found it): Would it be possible to add the MQTT retain flag in config file? The retain flag really helps to have your system in a reliable state after rebooting...

Best Regards
Michael

**Comment by jomjol** [comment]
Well. there should be a retain flag for the connect property. But you cannot mody it actively. What exactly do you need?

I'm not an expert in MQTT.

**Reply by Grillfreund** [reply]
Thank you for your reply,

The retain flag isn't set on connect. It's a property for the mqtt send command. It simply tells the server that the message as to be stored in mqtt server. If a client then subscribes to a topic, it immediately gets the last message and it hasn't to wait for a new message. Please see also https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/ for better explainaition.

**Comment by posi74de** [comment]
I'm also looking for this feature. Retain flag can be set on MQTT publish per message.

**Comment by gaszowc** [comment]
I join Grillfreund, a really cool project. I use it now half a year without problems. For me it is now getting more and more readings from different sensors. That's why I decided to use a MQTT-Broker as well. So that the database client does not have to constantly listen to a measured value, I want to run a cron job that fetches the last measured value from the broker every hour. For this the retain flag would be really a big help.
You have actually already prepared it: [https://github.com/jomjol/AI-on-the-edge-device/blob/master/code/components/jomjol_mqtt/interface_mqtt.cpp](https://github.com/jomjol/AI-on-the-edge-device/blob/master/code/components/jomjol_mqtt/interface_mqtt.cpp) row 22 and 55.
I would be happy if you could add it in the config.

Best regards

Hi, update.
I have adapted jomjol's project for my needs. Thanks for https://github.com/jomjol/AI-on-the-edge-device/wiki/Build-Instructions.
Retain is now true for me.

**Reply by caco3** [reply]
There where some changes to MQTT. AFAOK the retain flag gets set.
Maybe you could check?

How ever there are still some stability issues with MQTT, eg. it disconnects all the time.
